### DISCUSSION TITLE
sidekiq-ent unique lock for non-existing job

### DISCUSSION THREAD

**Discussion by jarl-dk** [sidekiq/sidekiq#6321] [Q&A] [answered]
I have a Job `AbcJob` which is configured 
```ruby
sidekiq_options unique_for: 10.minutes
```

When I do `AbcJob.perform_async` I get 
```
Skipping enqueue for AbcJob, not unique (lock held by JID 9c4160b4591146eb35f3b87a)
```
When I try to find this job, like
```ruby
jid = "9c4160b4591146eb35f3b87a"
Sidekiq::Queue.new('my_queue_name').find_job(jid)
Sidekiq::ScheduledSet.new.find_job(jid)
Sidekiq::RetrySet.new.find_job(jid)
Sidekiq::JobSet.new('my_queue_name').find_job(jid)
Sidekiq::DeadSet.new.find_job(jid)
``` 
I get no results. I believe that I have (intentionally) deleted this job earlier (maybe from the UI), but it seems like the lock is still there for a non-existing job.

Q1: Is it expected that a lock shall survive when a job is deleted?

Q2: How do I delete a lock for a no longer existing job?

Q3: How do I list all locks for jids?

**Comment by mperham** [comment]
1. Yes.
2. The Redis `DEL` command.
3. The Redis `SCAN` command with `SCAN 0 match unique:*`.

You're abusing the uniqueness feature, this isn't supported at all. The docs note that uniqueness longer than one day is a bad idea.

**Reply by jarl-dk** [reply]
OK. I have changed the original post to not let other irrelevant information distract focus away from the questions.

That said.
1. OK
2. Is that not possible using Sidekiq? Given a jid, what is the exact redis (ruby) commands to do this? I don't have redis-cli on the machine.
3. How do I do that from a ruby console?

4. Further, is it possible to have sidekiq raise an exception in the case of `Skipping enqueue for AbcJob, not unique` in addition or instead of the logging?

**Reply by mperham** [reply]
2. Is **what** not possible? Delete a lock?
3. `cl = RedisClient.new(url: ...).del(key)` where key is `unique:{hash}` and hash is `ctx = [klass, queue, args].join("|"); hash = Digest::SHA1.hexdigest(ctx)`. Alternatively `cl.scan("match", "unique:*", &:del)`
4. Nope.

**Reply by jarl-dk** [reply]
2. Yes. Is it possible to delete a lock for a no longer existing job (given a jid) using Sidekiq API? If so, what is the Sidekiq API command for that?
3. I don't see any `jid` in that? Is that is because the lock is not identified by the jid, but rather a unique context (klass, queue, args). I guess the last "alternative" way will delete **all** unique locks, right?

**Reply by mperham** [reply]
1. There is no Sidekiq API to delete a lock. There is an API to see if a lock exists. As I mentioned, you have to use SCAN and DEL to find and delete locks with the pattern "unique:*". Sidekiq does not expose this.

2. The locks are not keyed to a JID because how could two jobs share a unique lock if the lock is keyed to a specific JID? After Sidekiq Enterprise v7.0.4, unique locks will contain the JID as the value of the lock, i.e. "unique:123456789" => "jid1234567" so you can verify which job is holding the lock.

Please read the code in `lib/sidekiq-ent/unique.rb`.

**Reply by mperham** [reply, marked as answer]
Think of the JID as a primary key, the unique lock is a composite key of `(class, args, queue)`.
